src/csharp/Microsoft.Spark/Sql/RowCollector.cs

```diff
@@ -37,5 +38,81 @@ public IEnumerable<Row> Collect(ISocketWrapper socket)
                 }
             }
         }
+
+        /// <summary>
+        /// Synchronously collects pickled row objects from the given socket.
+        /// </summary>
+        /// <param name="socket">Socket the get the stream from</param>
+        /// <param name="server">The jvm socket auth server</param>
+        /// <returns>Collection of row objects</returns>
+        public IEnumerable<Row> SynchronousCollect(ISocketWrapper socket, JvmObjectReference server) =>
+            new SynchronousRowCollector(socket, server);
+
+        /// <summary>
+        /// SynchronousRowCollector synchronously collects Row objects from a socket.
+        /// </summary>
+        private class SynchronousRowCollector : IEnumerable<Row>
+        {
+            private readonly ISocketWrapper _socket;
+            private readonly JvmObjectReference _server;
+
+            private int _readStatus = 1;
+            private IEnumerable<Row> _collectEnumerable = null;
+
+            internal SynchronousRowCollector(ISocketWrapper socket, JvmObjectReference server)
+            {
+                _socket = socket;
+                _server = server;
+            }
+
+            ~SynchronousRowCollector()
+            {
+                // If iterator is not fully consumed
+                if ((_readStatus == 1) && (_collectEnumerable != null))
+                {
+                    // Finish consuming partition data stream
+                    foreach (Row _ in _collectEnumerable)
+                    {
+                    }
+
+                    // Tell Java to stop sending data and close connection
+                    Stream outputStream = _socket.OutputStream;
+                    SerDe.Write(outputStream, 0);
+                    outputStream.Flush();
+                }
+            }
+
+            public IEnumerator<Row> GetEnumerator()
+            {
+                Stream inputStream = _socket.InputStream;
+                Stream outputStream = _socket.OutputStream;
+
+                while (_readStatus == 1)
+                {
+                    // Request next partition data from Java
+                    SerDe.Write(outputStream, 1);
+                    outputStream.Flush();
+
+                    // If response is 1 then there is a partition to read, if 0 then fully consumed
+                    _readStatus = SerDe.ReadInt32(inputStream);
+                    if (_readStatus == 1)
+                    {
+                        // Load the partition data from stream and read each item
+                        _collectEnumerable = new RowCollector().Collect(_socket);
+                        foreach (Row row in _collectEnumerable)
+                        {
+                            yield return row;
+                        }
+                    }
+                    else if (_readStatus == -1)
+                    {
+                        // An error occurred, join serving thread and raise any exceptions from the JVM
```

Comment:
`raise any exceptions from the JVM` => does this mean if `getResult` is called, exception would be propagated?

Reply:
The exception will be thrown on the spark side and the exception stack trace will appear in the spark logs.  Do you think we should throw an exception in the dotnet side too ?

Reply:
Can you add that the exception stack trace will appear in the driver log to the comment?